File: ProyectoCivitasJava/src/civitas/TituloPropiedad.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;

public class TituloPropiedad {
    //Atributos de clase
    private static float factorInteresesHipoteca = 1.1f;
    
    //Atributos de instancia
    private float alquilerBase;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    
    //Atributos de referencia
    private Jugador propietario;
    
    TituloPropiedad(String nom, float ab, float fr, float hb, float pc, float pe){
        nombre = nom;
        alquilerBase = ab;
        factorInteresesHipoteca = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        propietario = null;
        numCasas = 0;
        numHoteles = 0;
        hipotecado = false;
    }
    
    void actualizaPropietarioPorConversion(Jugador jugador){
        propietario = jugador;
    }
    
    boolean cancelarHipoteca(Jugador jugador){
        boolean result = false;
        if (hipotecado){
            if(esEsteElPropietario(jugador)){
                jugador.paga(getImporteCancelarHipoteca());
                hipotecado = false;
                result = true;
            }
        }
        
        return result;
    }
    
    int cantidadCasasHoteles(){
        return numCasas+numHoteles;
    }
    
    boolean comprar(Jugador jugador){
        boolean result = false;
        if(!tienePropietario()){
            propietario = jugador;
            result = true;
            jugador.paga(getPrecioCompra());
            
        }
        return result;
    }
    
    boolean construirCasa(Jugador jugador){
        boolean result = false;
        if(esEsteElPropietario(jugador)){
            jugador.paga(precioEdificar);
            numCasas = numCasas + 1;
            result = true;
        }
        return result;
    }
    
    boolean construirHotel(Jugador jugador){
        boolean result = false;
        if(esEsteElPropietario(jugador)){
            jugador.paga(precioEdificar);
            numHoteles = numHoteles + 1;
            result = true;
        }
        return result;
    }
    
    boolean derruirCasas(int n, Jugador jugador){
        boolean comprobacion = true;
        if (jugador == propietario && numCasas >= n){
            numCasas -= n;
        }
        else
            comprobacion = false;
        return comprobacion;
    }
    
    private boolean esEsteElPropietario(Jugador jugador){
        boolean comprobacion = true;
        if (propietario != jugador){
            comprobacion = false;
        }
        
        return comprobacion;
    }
    
    public boolean getHipotecado(){
        return hipotecado;
    }
    
    float getImporteCancelarHipoteca(){
        return factorInteresesHipoteca*(getImporteHipoteca());
    }
    
    private float getImporteHipoteca(){
        return hipotecaBase*(1+(numCasas*0.5f)+(numHoteles*2.5f));
    }
    
    public String getNombre(){ //Debe ser publico para el método gestionar de vistatextual
        return nombre;
    }
    
    //visibilidad cambiada
    public int getNumCasas(){
        return numCasas;
    }
    
    
    //visibilidad cambiada
    public int getNumHoteles(){
        return numHoteles;
    }
    
    //visibilidad cambiada
    public float getPrecioAlquiler(){
        float alquiler;
        if(hipotecado || propietarioEncarcelado())
            alquiler = 0;
        else
            alquiler = alquilerBase*(1+(numCasas*0.5f)+(numHoteles*2.5f));
    
    return alquiler;
    }
    
    //cambiar visibilidad
    public float getPrecioCompra(){
        return precioCompra;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    private float getPrecioVenta(){
        return precioCompra+(numCasas+5*numHoteles)*precioEdificar*factorRevalorizacion;
    }
    
    //visibilidad cambiada
    public Jugador getPropietario(){
        return propietario;
    }
    
    boolean hipotecar(Jugador jugador){
        boolean salida = false;
        if ( !hipotecado && esEsteElPropietario(jugador)){
            propietario.recibe(getImporteHipoteca());
            hipotecado = true;
            salida = true;
        }
        return salida;
    }
    
    private boolean propietarioEncarcelado(){
        return propietario.isEncarcelado();
    }
    
    boolean tienePropietario(){
        boolean comprobacion = true;
        if (propietario == null){
            comprobacion = false;
        }
        return comprobacion;
    }
    
    @Override
    public String toString(){
        String salida;
        String nombreprop;
        
        if (propietario == null){
            nombreprop = "No tiene";
        }
        else{
            nombreprop = propietario.getNombre();
        }
        
        salida = "Precio de compra: " + precioCompra + " | Propietario: " + nombreprop + " | Alquiler Base: " + alquilerBase +
                 " | Número de casas: " + numCasas + " | Número de hoteles: " + numHoteles + " | ";
        return salida;
        
    }
    
    void tramitarAlquiler(Jugador jugador){
        if(!esEsteElPropietario(jugador) && tienePropietario()){
            float precio = getPrecioAlquiler();
            jugador.pagaAlquiler(precio);
            propietario.recibe(precio);
        }
        
    }
    
    boolean vender(Jugador jugador){
        boolean comprobacion = true;
        if(!hipotecado && esEsteElPropietario(jugador)){
            propietario.recibe(getPrecioVenta());
            propietario = null;
            numCasas = 0;
            numHoteles = 0;
        }
        else
            comprobacion = false;
        
        return comprobacion;
    }
    
    
    
}

File: ProyectoCivitasRuby/lib/modelo/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


module Civitas
  class TituloPropiedad
  
    #Atributo de clase
    @@factorintereseshipoteca = 1.1;
  
    #Constructor
    def initialize(nom, ab, fr, hb, pc, pe)
      #Atributos de instancia
      @nombre = nom
      @alquilerbase = ab
      @factorrevalorizacion = fr
      @hipotecabase = hb
      @preciocompra = pc
      @precioedificar = pe
      @numcasas = 0
      @numhoteles = 0
      @hipotecado = false
    
      #Atributo de referencia
      @propietario = nil
    end
  
    #Lectores
    attr_reader :hipotecado, :propietario, :precioedificar, :preciocompra, :numhoteles, :numcasas, :nombre

    
    def actualiza_propietario_por_conversion(jugador)
      @propietario = jugador
    end
  
    
    def cancelar_hipoteca(jugador)
      result = false
      if(@hipotecado)
        if(es_este_el_propietario(jugador))
          @propietario.paga(get_importe_cancelar_hipoteca())
          @hipotecado = false
          result = true
        end
      end
      result
    end
    
  
    def cantidad_casas_hoteles
      @numcasas + @numhoteles
    end
  
    
    def comprar(jugador)
      result = false
      if(tiene_propietario == false)
        actualiza_propietario_por_conversion(jugador)
        result = true
        @propietario.paga(@preciocompra)
      end
      result
    end
    
  
    def construir_casa(jugador)
      result = false
      if (es_este_el_propietario(jugador))
        jugador.paga(@precioedificar)
        @numcasas = @numcasas+1
        result = true
      end
      result
    end
    
  
    def construir_hotel(jugador)
      result = false
      if (es_este_el_propietario(jugador))
        jugador.paga(@precioedificar)
        @numhoteles = @numhoteles+1
        result = true
      end
      result
    end
  
  
    def derruir_casas(n, jugador)
      comprobacion = false
      if(es_este_el_propietario(jugador) && @numcasas >= n)
        @numcasas=@numcasas-n
        comprobacion = true
      end
      comprobacion
    end
  
    
    def get_importe_cancelar_hipoteca()
      @@factorintereseshipoteca*get_importe_hipoteca()
    end
  
    
    def hipotecar(jugador)
      salida = false
      if (!@hipotecado && es_este_el_propietario(jugador))
        @propietario.recibe(get_importe_hipoteca)
        @hipotecado = true
        salida = true
      end
      salida
    end
  
    
    def tiene_propietario
      @propietario!=nil
    end
 
    
    def tramitar_alquiler(jugador)
      if(tiene_propietario && !es_este_el_propietario(jugador))
        precio = get_precio_alquiler()
        jugador.paga_alquiler(precio)
        @propietario.recibe(precio);
      end
    end
  
    
    def vender(jugador)
      comprobacion = false
      if(es_este_el_propietario(jugador) && !hipotecado)
        cantidad = get_precio_venta
        @propietario.recibe(cantidad)
        actualiza_propietario_por_conversion(nil)
        @numcasas = 0
        @numhoteles = 0
        comprobacion = true
      end
      comprobacion
    end
  

    def to_s
      if (@propietario == nil)
        nombreprop = "no tiene"
      else
        nombreprop = @propietario.nombre
      end
      salida = "Alquiler: #{@alquilerbase} | Propietario: #{nombreprop} | Precio: #{@preciocompra} | Numero de casas: #{@numcasas} | Numero de hoteles: #{@numhoteles}"
    end

    
    
    private
    
    def propietario_encarcelado
      comprobacion = false
      if (@propietario.isencarcelado && tiene_propietario)
        comprobacion = true
      end
      comprobacion
    end
  

    def get_precio_venta
      @preciocompra+(@numcasas+5*@numhoteles)*@precioedificar*@factorrevalorizacion
    end
  
    
    def get_precio_alquiler
      if(@hipotecado || propietario_encarcelado)
        cantidad = 0
      else
        cantidad = @alquilerbase*(1+(@numcasas*0.5)+(@numhoteles*2.5))
      end
      cantidad
    end
  
    
    def es_este_el_propietario(jugador)
      @propietario==jugador
    end

    
    def get_importe_hipoteca
      @hipotecabase*(1+(@numcasas*0.5)+(@numhoteles*2.5))
    end
    
  end
end
